File: gui/simlab/src/types/milp.ts
// MILP model catalog and parameter sweeps (mirrors rest-api/api/domain/milp.py)

export interface MilpParamSpecDto {
  name: string;
  description: string;
  default: number;
  sweepable: boolean;
}

export interface MilpModelInfoDto {
  key: string;
  problem_key: string;
  title: string;
  description: string;
}

export interface MilpModelDto extends MilpModelInfoDto {
  parameters: MilpParamSpecDto[];
  solver_defaults: Record<string, number>;
  formulation: string;
}

export interface MilpSolverConfigDto {
  backend: string;
  time_limit_s: number;
  mip_gap: number;
  allow_fallback: boolean;
}

export interface MilpSweepProgressDto {
  total_combos: number;
  done: number;
  solved: number;
  infeasible: number;
  unique_genotypes: number;
}

export interface MilpSweepInfoDto {
  id: string;
  name: string;
  model_key: string;
  status: string | null;
  system_message: string | null;
  created_time: string | null;
  start_time: string | null;
  end_time: string | null;
  progress: MilpSweepProgressDto;
  experiment_id: string | null;
  campaign_id: string | null;
  cancel_requested: boolean;
}

export interface MilpSolutionRecord {
  index: number;
  params: Record<string, number>;
  status: string;
  genotype: string | null;
  mask: number[] | null;
  n_installed: number | null;
  obj_value: number | null;
  mip_gap: number | null;
  runtime_s: number;
  is_duplicate: boolean;
  message: string;
}

export interface MilpSweepDto extends MilpSweepInfoDto {
  problem: Record<string, unknown>;
  problem_id: string | null;
  parameter_grid: Record<string, number[]>;
  fixed_parameters: Record<string, number>;
  solver: MilpSolverConfigDto;
  batch_options: Record<string, unknown>;
  solutions: MilpSolutionRecord[];
}

export interface MilpSweepCreatePayload {
  name: string;
  model_key: string;
  problem: Record<string, unknown>;
  problem_id?: string | null;
  parameter_grid: Record<string, number[]>;
  fixed_parameters: Record<string, number>;
  solver: MilpSolverConfigDto;
  batch_options: Record<string, unknown>;
  campaign_id?: string | null;
}

export interface MilpEngineStatusDto {
  status: string; // online | offline | unknown
  solver: string | null;
  gurobi_license: string | null;
  available_backends: string[];
  updated_time: string | null;
}

File: rest-api/api/domain/milp.py
from typing import Any, Optional
from pydantic import BaseModel, Field


# ---------------------------------------------------------------- model catalog

class MilpParamSpecDto(BaseModel):
    name: str
    description: str
    default: float
    sweepable: bool


class MilpModelInfoDto(BaseModel):
    key: str
    problem_key: str
    title: str
    description: str


class MilpModelDto(MilpModelInfoDto):
    parameters: list[MilpParamSpecDto]
    solver_defaults: dict[str, float]
    formulation: str


# --------------------------------------------------------------------- sweeps

class MilpSolverConfigDto(BaseModel):
    backend: str = "gurobi"
    time_limit_s: float = 300.0
    mip_gap: float = 0.01
    # When the requested backend is unavailable (e.g. no Gurobi license),
    # let the engine fall back to another available backend instead of failing.
    allow_fallback: bool = True


class MilpSweepCreateDto(BaseModel):
    name: str
    model_key: str
    # Exported problem JSON — same shape the GUI sends in experiment
    # parameters.problem (problem drafts are a GUI-internal format).
    problem: dict[str, Any]
    problem_id: Optional[str] = None  # provenance: problems collection draft
    parameter_grid: dict[str, list[float]]
    fixed_parameters: dict[str, float] = Field(default_factory=dict)
    solver: MilpSolverConfigDto = Field(default_factory=MilpSolverConfigDto)
    # Passed through to the generated batch experiment (simulation config,
    # objectives, source_repository_options, data_conversion_config, ...).
    batch_options: dict[str, Any] = Field(default_factory=dict)
    campaign_id: Optional[str] = None


class MilpSweepProgressDto(BaseModel):
    total_combos: int = 0
    done: int = 0
    solved: int = 0
    infeasible: int = 0
    unique_genotypes: int = 0


class MilpSweepInfoDto(BaseModel):
    id: str
    name: str
    model_key: str
    status: Optional[str] = None
    system_message: Optional[str] = None
    created_time: Optional[str] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    progress: MilpSweepProgressDto = Field(default_factory=MilpSweepProgressDto)
    experiment_id: Optional[str] = None
    campaign_id: Optional[str] = None
    cancel_requested: bool = False


class MilpSweepDto(MilpSweepInfoDto):
    problem: dict[str, Any] = Field(default_factory=dict)
    problem_id: Optional[str] = None
    parameter_grid: dict[str, list[float]] = Field(default_factory=dict)
    fixed_parameters: dict[str, float] = Field(default_factory=dict)
    solver: MilpSolverConfigDto = Field(default_factory=MilpSolverConfigDto)
    batch_options: dict[str, Any] = Field(default_factory=dict)
    # One entry per solved grid combination (params, status, genotype, ...)
    solutions: list[dict[str, Any]] = Field(default_factory=list)


class MilpEngineStatusDto(BaseModel):
    status: str = "unknown"  # online | offline | unknown
    solver: Optional[str] = None
    gurobi_license: Optional[str] = None
    available_backends: list[str] = Field(default_factory=list)
    updated_time: Optional[str] = None
